lib/property_constraints.js

```diff
@@ -1,95 +1,116 @@
 var Immutable = require('immutable');
 
-var cardinality = {
-  '=': function(val, count) {
-    return val.size === parseInt(count);
-  },
-  '>': function(val, count) {
-    return val.size > parseInt(count);
-  },
-  '<': function(val, count) {
-    return val.size < parseInt(count);
-  }
+function not(fn) {
+  return function(state) {
+    return !fn(state);
+  };
+}
+
+function and() {
+  var fns = Array.prototype.slice.call(arguments);
+  return function(state) {
+    return fns.reduce(function(memo, fn) {
+      return memo && fn(state);
+    }, true);
+  };
+}
+
+
+var constraint = function() {
+  var fns = Array.prototype.slice.call(arguments);
+  return function(state) {
+    return fns.reduce(function(previousValue, fn) {
+      return (!!previousValue) && fn(previousValue);
+    }, state);
+  };
 };
 
-var handleFunctionParameter = function(expectedSet, propertyValue) {
-  if (typeof (expectedSet) === 'function' && Immutable.Iterable.isIterable(propertyValue)) {
-    return propertyValue.filter(expectedSet);
-  }
-  return expectedSet;
+
+var hasProperty = function(propertyName) {
+  return function(state) {
+    return state.hasOwnProperty(propertyName) && !!state[propertyName];
+  };
 };
 
-var rules = {
+var get = function(attrName) {
+  return function(state) {
+    return state[attrName];
+  };
+};
 
-  hasProperty: function(model, propertyName) {
-    return model.get(propertyName);
-  },
+var getIn = function() {
```

Comment:
How is Immutable exposed to the client? But I do see the coercion problem. I felt like hand-rolling deep object retrieval was not that easy. Essentially I was hoping that we might want to set a particular object hierarchy as root(sort of like CWD), then have transition rules based on this new context root.